### DISCUSSION TITLE
how to programmatically select item(s) in selection list

### DISCUSSION THREAD

**Discussion by haggletonpie** [MyreMylar/pygame_gui#218] [General]
(thanks for making this library!)

I need to 'select' (and maybe 'hover') items in a selection list via code, not just with mouse clicks.

Suggestions?

1st example: Single select: the selection list box responds to keypress events to scroll up/down thru the items, and the item under the 'scroll cursor' is 'selected'.

2nd example: Multi select: the selection list box responds to keypress events to scroll thru the items, 'hovering' the current cursor position, then responds to another keypress to select/deselect that item.

Thanks much.

**Comment by MyreMylar** [comment]
I expect you could manage this with a lot of inheriting and overriding methods. The UISelectionList is made up of UIButtons, so you'd need to subclass both of those and then override the `process_events()` methods of both to switch out button clicks for button presses and override the methods where UISelectionList creates UIButtons.

It'd likely be a fair chunk of work.

You could also fake the mouse events using:

https://www.pygame.org/docs/ref/event.html#pygame.event.post

This is what I do for testing.

There is a plan to eventually add some keyboard (and from there, controller buttons) controls to the overall UI, much as you can use Tab and Return to move around and activate a highlight on most normally mouse driven GUIs.

**Comment by haggletonpie** [comment]
>It'd likely be a fair chunk of work.

"Actually, it's super easy, barely an inconvenience!", at least for what I needed.

I just used your testing code to simulate a click/selection/deselection....

https://github.com/MyreMylar/pygame_gui/blob/0cbf7056518377b455d51a8d20167f4029756ad9/tests/test_elements/test_ui_selection_list.py#L32
```
        event_data = {'user_type': pygame_gui.UI_BUTTON_PRESSED,
                      'ui_element': selection_list.item_list_container.elements[0]}
        press_list_item_event = pygame.event.Event(pygame.USEREVENT, event_data)
        default_ui_manager.process_events(press_list_item_event)
```

As for the simulating hovering, and 'keyboard-ifying' the UI, yeah, I can how that'd take some time.

Thx for the reply and the library.

**Comment by haggletonpie** [comment]
> >  'hover' items in a selection list via code, not just with mouse clicks.

>  It'd likely be a fair chunk of work.

I can't help but think there's a fairly easy way to do this. Some questions ...

> You could also fake the mouse events using:
> https://www.pygame.org/docs/ref/event.html#pygame.event.post

Q: Since it was so easy to use that method to post a click event, couldn't a hover/unhover event be posted to an item in a selection list? If so, how would I express that to pygame_gui?

Q: Via code (not a .json theme file), can we set individual theme properties for individual items in a selection list? Or at least dynamically assign a theme declaration to an individual item in a selection list?

What I'm implementing here is a multi-selection-list that gains focus via the keyboard, and then, with cursor keys, the user can scroll up and down to highlight, then select/deselect items in the selection list...

... which means that there could be 4 themes declared in the json file: normal-notselected, normal-selected, highlighted-notselected, highlighted-selected

However, if theme properties can be set directly in code, it seems that would be easier...

... especially if font properties of the items can be set directly/dynamically, which would mean that a highlighted item can simply be changed to a larger font size, or bold/italic.

**Comment by MyreMylar** [comment]
Lots of nested questions here.

1. Right now 'hovering' is implemented at a UIManager wide level, this is to manage 'unhovering' so that when your mouse is no longer over a thing we make sure that the thing knows it is no longer being hovered. When this means is that if you want to override this behaviour to have the keyboard command 'hovering' the update function of UIManger would have to be modified. I think eventually the library would want something that disabled & enabled the call: `ui_element.check_hover()` based on whether the last interaction came from a mouse movement, a keyboard navigation key or a gamepad stick/d-pad movement to support all of those flexibly. So while you could post some hover related events, or call some hover functions right now - they would be counteracted immediately by the existence of the mouse in whatever position it was currently sitting for your app.
2. Right now colours are easy for different button states, but changing fonts is not. Generally this is because fonts change the shape of things and are slower to deal with. Colours (even font colours) are basically free because we are redrawing constantly anyway and which colour we use makes no difference. Fonts require loading from the file when you change anything about them which the developer/user of pygame GUI may or may not have done. In the normal case of bolding a font when you hover a button you could be waiting up to a two seconds the first time you hover a button in that style while the font loads from disc in the bold version. There is also problems with fonts taking up different amounts of space (where changing colours does not). Now for a simple button we could just ignore this, it might not be noticeable when the size change is small anyway (e.g. from normal to bold). However, then supporting this for simple button cases could cause a ripple effect of issues with longer strings of text, with multi-line text and with relative UI layouts. Anyway that is to explain why there are state related theming options of colours but not for fonts.
3. To change thing stuff programmatically outside of the theme file is generally just a case of just doing the process that happens when you edit the theme file while the app is running except with you changing the values in code rather than in the theme file. For the theme file with live updates enabled we check whether the theme file has been edited and reload it if so. Then if it has we call `rebuild_from_changed_theme_data()` on everything in the UI. So if you just wanted to change the font on a button at a certain point you could first edit the theme data for that button:

`my_button.ui_theme.ele_font_res[my_button.combined_element_ids[0]] = new_font`

then call `rebuild_from_changed_theme_data()`

as an example. So long as you don't trigger a reload of the theme file itself the data in the them in memory will be whatever you set for that button's ID. You could load your new font in the standard way or use the pygame gui font dictionary.

I hope some of that helps. Apologies there has not been much movement on keyboard navigation yet. It's been a slow year for open source stuff for me due to still ongoing COVID pandemic increasers to my the work part of my work/fun code/life balance.

**Comment by haggletonpie** [comment]
thx much for the help and the library. will post my results.

**Comment by haggletonpie** [comment]
The complexity of how to get/set theming properties via code is beyond my comprehension without docs and examples.

Would you mind providing an example? Here's the code I'm using to fire an event...

```
the_element = thelistui.item_list_container.elements[index] 
event_data = {'user_type': pygame_gui.UI_BUTTON_PRESSED,
                'ui_element': the_element}
press_list_item_event = pygame.event.Event(pygame.USEREVENT, event_data)

```

Could you write out how exactly to get and set the font color for `the_element`?

**Comment by MyreMylar** [comment]
> The complexity of how to get/set theming properties via code is beyond my comprehension without docs and examples.
> 
> Would you mind providing an example? Here's the code I'm using to fire an event...
> 
> ```
> the_element = thelistui.item_list_container.elements[index] 
> event_data = {'user_type': pygame_gui.UI_BUTTON_PRESSED,
>                 'ui_element': the_element}
> press_list_item_event = pygame.event.Event(pygame.USEREVENT, event_data)
> ```
> 
> Could you write out how exactly to get and set the font color for `the_element`?


Well the font colour specifically is easiest to set with a theme file so it is all setup ahead of time, but outside of that... the snippet I posted above should set you on the right path:

`my_button.ui_theme.ele_font_res[my_button.combined_element_ids[0]] = new_font`

in your case above the element will be a button and you want to change a colour, not a font so you can switch it to:

```
the_element.ui_theme.ui_element_colours[the_element.combined_element_ids[0]]['selected_text'] = pygame.Color('#ff0000ff')
the_element.rebuild_from_changed_theme_data()
```

if you want to change the colour of the text after selection (the states for text colour are called 'normal_text', 'hovered_text', 'disabled_text', 'selected_text', 'active_text').

What we are doing here is editing the data for the theme file that exists in memory after the theme file is loaded (or technically after the default theme file and any custom theme files are loaded). so the first `.` after `the_element` grabs the element's reference to the single UI theme that exists, the second `.` narrows it down to the data structure in the theme file ( `ui_element_colours` - note the 'u' in colour here because I'm British) that holds element's colour information. The data structure that hold colour information is a dictionary of dictionaries. 

Elements have a list of IDs preceding from most specific to least specific (this is the `.combined_element_ids` part), in this case because we only want to mess with this particular element we set the most specific ID available (at the 0th index). Depending on how you set up your selection list (with strings or tuples for the items and whether you setup an ID for the list) this may set the theming for every item in every selection list, or just this item in this specific list.

I haven't tested this setup, this description of how it works is from memory and looking at the code afresh this morning - you may encounter some unforeseen issues in practice but the basic thrust of it should be sound.

**Comment by wadsworj** [comment]
I would add my vote to implementing "programmatically select item(s) in selection list".
